rustls/src/msgs/codec.rs

```diff
@@ -85,6 +85,68 @@ impl<'a> Reader<'a> {
     }
 }
 
+/// A version of [`Reader`] that operates on mutable slices
+pub(crate) struct ReaderMut<'a> {
+    /// The underlying buffer storing the readers content
+    buffer: &'a mut [u8],
+    used: usize,
+}
+
+#[allow(dead_code)]
+impl<'a> ReaderMut<'a> {
+    pub(crate) fn init(bytes: &'a mut [u8]) -> Self {
+        Self {
+            buffer: bytes,
+            used: 0,
+        }
+    }
+
+    pub(crate) fn as_reader<T>(&mut self, f: impl FnOnce(&mut Reader) -> T) -> T {
+        let mut r = Reader {
+            buffer: self.buffer,
+            cursor: 0,
+        };
+        let ret = f(&mut r);
+        let cursor = r.cursor;
+        self.used += cursor;
+        let (_used, rest) = mem::take(&mut self.buffer).split_at_mut(cursor);
+        self.buffer = rest;
+
+        ret
+    }
+
+    pub(crate) fn rest(&mut self) -> &'a mut [u8] {
+        let rest = mem::take(&mut self.buffer);
+        self.used += rest.len();
+        rest
+    }
+
+    pub(crate) fn sub(&mut self, length: usize) -> Result<Self, InvalidMessage> {
+        match self.take(length) {
+            Some(bytes) => Ok(ReaderMut::init(bytes)),
+            None => Err(InvalidMessage::MessageTooShort),
+        }
+    }
+
+    pub(crate) fn used(&self) -> usize {
+        self.used
+    }
+
+    fn left(&self) -> usize {
+        self.buffer.len()
+    }
+
+    fn take(&mut self, length: usize) -> Option<&'a mut [u8]> {
+        if self.left() < length {
+            return None;
+        }
+        let (taken, rest) = mem::take(&mut self.buffer).split_at_mut(length);
+        self.used += taken.len();
+        self.buffer = rest;
+        Some(taken)
+    }
```

Comment:
Nit: I feel like these should remain in the same order as for `Reader`?

Reply:
Fixed.